teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+  * Pro: This will improve the reusability of Tekton components by enabling the scenario of a Task providing an array result
+    and then a Pipeline being able to loop over those value for subsequent Tasks (i.e. reusing Tasks which are designed
+    for single inputs, but users want to use 'for each' item in an array)
+* [Simplicity](https://github.com/tektoncd/community/blob/main/design-principles.md#simplicity)
+  * Pro: This proposal reuses the existing array or string concept for params
+  * Pro: This proposal continues [the precedent of using JSONPath syntax in variable replacement](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-561476075)
+  * Con: There are now two different files/paths to which users can write results and they have to understand when
+    to use which. (We could choose to deprecate the original path if necessary.)
+* [Flexibility](https://github.com/tektoncd/community/blob/main/design-principles.md#flexibility)
+  * Con: Although there is a precedent for including JSONPath syntax, this is a step toward including more hard coded
+    expression syntax in the Pipelines API (without the ability to choose other language options)
+* [Conformance](https://github.com/tektoncd/community/blob/main/design-principles.md#conformance)
+  * Supporting array results and indexing syntax would be included in the conformance surface
+
+## Drawbacks
+
+* Increases the size of TaskRuns (by including arrays in termination messages and results)
+* Only partial support for JSONPath (i.e. just array indexing support, nothing more)
+
+## Alternatives
+
+1. Do not support arrays explicitly - if we do this, we probably shouldn't do TEP-0075 either, i.e. results should
+  always just be strings
+1. Add support for array results, but don't add support for indexing into arrays
+   * Con: In TEP-0075 we propose adding support for dictionaries; if we take the same approach as this proposal, this
+     would mean not allowing the ability to get individual values out of those dictionaries which would significantly
+     limit the number of Tasks this would be compatible with (i.e. we should be consistent between arrays and
+     dictionaries)
+   * Con: Without indexing, arrays will only work with arrays. This means you cannot combine arrays with Tasks that
+     only take string parameters and this limits the interoperabiltiy between Tasks.
+1. Instead of adding support for a `.json` file in addition to the existing path, we use the same file and parse it as
+  json.
+    * Pro: Avoids the complication of two different paths for result files
+    * Con: A backwards incompatible change for Tasks which are currently writing json to the results file and treating
+      it as a string. The feature would be behind an alpha flag so they wouldn't be impacted immediately but eventually
+      they would need to update their Tasks to escape the json.
```

Comment:
This is definitely a valid concern, not really highlighted above I think. What happens to those use-cases once we rely solely on the json syntax for results ? How a user can pass a json as a string result ?

Reply:
the original proposal was that users when users wanted to emit something other than a string, they'd write it to the `.json` file - otherwise anything they write to the results file would be treated as just a string so they wouldn't need to change anything (i.e. we'd never rely soley on json syntax for results - we'd support both the current "string" syntax and the json syntax)

@pierretasci and @sbwsg pointed out an improvement we could make where we can (have our 🍰  and eat it to) support both without needing to introduce a new results file - we'd rely on how the Task defines the result in its spec to tell us if what we encounter in the file should be treated as json or treated as a string)